### DISCUSSION TITLE
Streaming Response from another API Service

### DISCUSSION THREAD

**Discussion by jimcarreer** [fastapi/fastapi#7754] [Questions] [answered]
### Description
Can an API be created / hooked up to stream the response from a request made to an upstream service?  I've seen this done in Node frameworks many times, and wondering how I might go about it with python + FastAPI.

### Additional context
A common architecture in microservices on K8 is to have a gateway layer that implements a "public" API that largely calls to various "backend services" that are internal to the cluster.  The gateway service is the only "exposed" service in this setup:
```

                            ================= K8 Network ===============
                            ||                                        ||
                            ||           -------> Backend Service A   ||
                            ||          /                             ||
                            ||         /                              ||
External Request ===> Gateway Service * --------> Backend Service B   ||
                            ||         \                              ||
                            ||          \                             ||
                            ||           -------> Backend Service C   ||
                            ||                                        ||
                            ============================================
```
So in this scenario, I'm trying to implement "Gateway Service" with FastAPI, and basically it has two use cases:
1) One where it must call several back end services to fulfill the external request and build the proper response.  **I don't need help with this, I already know how to do it**
2) Where Gateway Service is acting **almost** like a proxy / nginx, the only thing it needs to do is stream the response directly from the backend service.

Number 2 is kind of where I'm looking for guidance / examples.

**Comment by tiangolo** [comment, marked as answer]
@jimcarreer you can use Starlette's [`StreamingResponse`](https://www.starlette.io/responses/#streamingresponse). It takes something that can be iterated, `async` or normal iterable (if normal iterable, it's iterated on a threadpool, assuming it could be blocking).

You can pass to that anything that can be iterated, like a file, generator, etc.

**Comment by jimcarreer** [comment]
Excellent Ill look into using that for my project. Thanks! If I run into trouble Ill come back to this issue with questions.

**Comment by mateusz-glowinski** [comment]
There is example of potential solution: https://github.com/tiangolo/fastapi/issues/1788#issuecomment-1071222163
